### PR TITLE
Support expr with json field

### PR DESCRIPTION
**Pull request by xiaocai2333**
/kind improvement

### CODE REVIEW

File: tests/integration/json_expr_test.go [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+func doSearch(cluster *MiniCluster, collectionName string, outputField []string, expr string, dim int, t *testing.T) {
+	nq := 1
+	topk := 10
+	roundDecimal := -1
+
+	params := getSearchParams(IndexFaissIvfFlat, distance.L2)
+	searchReq := constructSearchRequest("", collectionName, expr,
+		floatVecField, schemapb.DataType_FloatVector, outputField, distance.L2, params, nq, dim, topk, roundDecimal)
+
+	searchResult, err := cluster.proxy.Search(context.Background(), searchReq)
```
Comment (xiaofan-luan) [inline tests/integration/json_expr_test.go:313]: need the check the search result here

File: tests/integration/json_expr_test.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+	"github.com/milvus-io/milvus-proto/go-api/commonpb"
+	"github.com/milvus-io/milvus-proto/go-api/milvuspb"
+	"github.com/milvus-io/milvus-proto/go-api/schemapb"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/funcutil"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/zap"
+)
+
+func TestJsonExpr(t *testing.T) {
```
Comment (xiaofan-luan) [inline tests/integration/json_expr_test.go:40]: Add test with invalid json?
For example nested json?

**Comment by longjiquan** [comment]
It will be really nice if you can add some examples in the pr body. With the example, it's much easier for us to understand the code changes. @xiaocai2333

File: internal/parser/planparserv2/Plan.g4 [hunk 3]
```diff
@@ -16,15 +17,16 @@ expr:
 	| expr op = (SHL | SHR) expr							                # Shift
 	| expr op = (IN | NIN) ('[' expr (',' expr)* ','? ']')                  # Term
 	| expr op = (IN | NIN) EmptyTerm                                        # EmptyTerm
-	| expr op1 = (LT | LE) Identifier op2 = (LT | LE) expr		            # Range
-	| expr op1 = (GT | GE) Identifier op2 = (GT | GE) expr                  # ReverseRange
+	| expr op1 = (LT | LE) (Identifier | JsonFiled) op2 = (LT | LE) expr	# Range
+	| expr op1 = (GT | GE) (Identifier | JsonFiled) op2 = (GT | GE) expr    # ReverseRange
```
Comment (longjiquan) [inline internal/parser/planparserv2/Plan.g4:21]: This expression should not be supported.

File: internal/parser/planparserv2/parser_visitor.go [hunk 4]
```diff
@@ -531,16 +543,22 @@ func (v *ParserVisitor) VisitEmptyTerm(ctx *parser.EmptyTermContext) interface{}
...
 	}
 
-	columnInfo := toColumnInfo(childExpr)
-	if columnInfo == nil {
+	return v.getColumnInfoFromJSON(child.GetText())
+}
+
+// VisitRange translates expr to range plan.
+func (v *ParserVisitor) VisitRange(ctx *parser.RangeContext) interface{} {
+	columnInfo, err := v.getChildColumnInfo(ctx.Identifier(), ctx.JsonFiled())
```
Comment (longjiquan) [inline internal/parser/planparserv2/parser_visitor.go:560]: Binary range with json should not be supported.

**Review by longjiquan — CHANGES_REQUESTED** [review]

**Comment by xiaocai2333** [comment]
> It will be really nice if you can add some examples in the pr body. With the example, it's much easier for us to understand the code changes. @xiaocai2333

tests/integration/json_expr_test.go

File: internal/core/src/query/ExprImpl.h [hunk 5]
```diff
@@ -30,10 +30,10 @@ template <typename T>
 struct TermExprImpl : TermExpr {
     const std::vector<T> terms_;
 
-    TermExprImpl(const FieldId field_id,
-                 const DataType data_type,
-                 const std::vector<T>& terms)
-        : TermExpr(field_id, data_type), terms_(terms) {
+    TermExprImpl(ColumnInfo column,
+                 const std::vector<T>& terms,
+                 const proto::plan::GenericValue::ValCase val_case)
```
[the remark this answers is not in the document]
Comment (xiaocai2333) [reply internal/core/src/query/ExprImpl.h:35]: data type is `JSON`, but `val_case` maybe int or string.

File: internal/core/src/query/visitors/ExecExprVisitor.cpp [hunk 6]
```diff
@@ -753,23 +873,32 @@ ExecExprVisitor::ExecBinaryRangeVisitorDispatcherJson(BinaryRangeExpr& expr_raw)
             expr.column_.field_id, index_func, elem_func);
     } else if (lower_inclusive && !upper_inclusive) {
         auto elem_func = [&](const milvus::Json& json) {
-            auto x = json.template at<GetType>(expr.column_.nested_path);
```
[the remark this answers is not in the document]
Comment (xiaocai2333) [reply internal/core/src/query/visitors/ExecExprVisitor.cpp:756]: why?

File: tests/integration/json_expr_test.go (hunk 2 above)
Comment (xiaocai2333) [reply tests/integration/json_expr_test.go:41]: add it

File: internal/parser/planparserv2/plan_parser_v2_test.go [hunk 7]
```diff
@@ -55,7 +58,7 @@ func TestExpr_Term(t *testing.T) {
 
 	exprStrs := []string{
 		`BoolField in [true, false]`,
-		`Int8Field in [1, 2]`,
```
Comment (czs007) [inline internal/parser/planparserv2/plan_parser_v2_test.go:58]: just keep this case

Comment (xiaocai2333) [reply internal/parser/planparserv2/plan_parser_v2_test.go]: done

File: internal/parser/planparserv2/utils.go [hunk 8]
```diff
@@ -307,7 +320,7 @@ func handleCompare(op planpb.OpType, left *ExprWithType, right *ExprWithType) (*
 }
 
 func relationalCompatible(t1, t2 schemapb.DataType) bool {
-	both := typeutil.IsStringType(t1) && typeutil.IsStringType(t2)
+	both := (typeutil.IsStringType(t1) || typeutil.IsJSONType(t1)) && (typeutil.IsStringType(t2) || typeutil.IsJSONType(t2))
```
Comment (xiaofan-luan) [inline internal/parser/planparserv2/utils.go:323]: questions? what is the relationalCompatible mean?
Json can be compared with string or not?
Json with int?
String with int?

**Pull request merged**